minerl/Malmo/Minecraft/src/main/java/com/microsoft/Malmo/Utils/MineRLTypeHelper.java

```diff
@@ -0,0 +1,75 @@
+package com.microsoft.Malmo.Utils;
+
+import com.google.gson.JsonObject;
+import net.minecraft.entity.player.InventoryPlayer;
+import net.minecraft.item.Item;
+import net.minecraft.item.ItemStack;
+import net.minecraft.util.ResourceLocation;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+public class MineRLTypeHelper {
+
+
+    /**
+     * @return A String item type suitable for Types.xsd or passing back to MineRL in messages, or
+     *     null if such a String could not be found.
+     */
+    @Nullable
+    public static String getItemType(Item item) {
+        ResourceLocation resourceLocation = Item.REGISTRY.getNameForObject(item);
+        if (resourceLocation == null) {
+            return null;
+        } else {
+            return resourceLocation.getResourcePath();
+        }
+    }
+
+    @NotNull
+    public static JsonObject jsonFromItemStack(ItemStack stack) {
+        JsonObject jobj = new JsonObject();
+        writeItemStackToJson(stack, jobj);
+        return jobj;
+    }
+
+    public static void writeItemStackToJson(ItemStack stack, JsonObject jsonObject) {
+        int metadata = stack.getMetadata();
+        if (metadata < 0 || metadata > 15) {
+            throw new RuntimeException(String.format("Unexpected metadata value %d.", metadata));
+        }
+        jsonObject.addProperty("type", MineRLTypeHelper.getItemType(stack.getItem()));
+        jsonObject.addProperty("metadata", metadata);
+        jsonObject.addProperty("quantity", stack.getCount());
+    }
+
+    /**
+     * Find the lowest inventory index that matches the itemType and the metadata.
+     * @param itemType The String name of the Item or ItemStack.
+     * @param metadata The metadata or the damage of the ItemStack. Pass null to ignore this constraint.
+     * @return The lowest inventory index matching the contraints, or null if no such slot is found.
+     */
+    @Nullable
+    public static Integer inventoryIndexOf(InventoryPlayer inventory, String itemType, @Nullable Integer metadata) {
+        Item parsedItem = Item.getByNameOrId(itemType);
+        if (parsedItem == null || parsedItem.getRegistryName() == null)
+            throw new IllegalArgumentException(itemType);
+        ResourceLocation targetRegName = parsedItem.getRegistryName();
+
+        for (int i = 0; i < inventory.getSizeInventory(); i++) {
+            ItemStack stack = inventory.getStackInSlot(i);
+            ResourceLocation regName = stack.getItem().getRegistryName();
+            if (regName == null) {
+                // 80% confident this will be AIR if the slot is empty, rather than null.
```

Comment:
Returns null if there is no such name in registry

Reply:
>public final ResourceLocation getRegistryName()
        {
            if (delegate.name() != null) return delegate.name();
            return registryName != null ? registryName : null;
        }

Reply:
>    Class<V> getRegistryType();

    // Default implementation, modders who make extra items SHOULD extend this instead of Object.
    // We have to do this until we get default implementations in Java 8.
    @SuppressWarnings({ "serial", "unchecked" })
    public static class Impl<T  extends IForgeRegistryEntry<T>> implements IForgeRegistryEntry<T>
    {
        private TypeToken<T> token = new TypeToken<T>(getClass()){};
        public final RegistryDelegate<T> delegate = PersistentRegistryManager.makeDelegate((T)this, (Class<T>)token.getRawType());
        private ResourceLocation registryName = null;

        public final T setRegistryName(String name)
        {
            if (getRegistryName() != null)
                throw new IllegalStateException("Attempted to set registry name with existing registry name! New: " + name + " Old: " + getRegistryName());

            int index = name.lastIndexOf(':');
            String oldPrefix = index == -1 ? "" : name.substring(0, index);
            name = index == -1 ? name : name.substring(index + 1);
            ModContainer mc = Loader.instance().activeModContainer();
            String prefix = mc == null || (mc instanceof InjectedModContainer && ((InjectedModContainer)mc).wrappedContainer instanceof FMLContainer) ? "minecraft" : mc.getModId().toLowerCase();
            if (!oldPrefix.equals(prefix) && oldPrefix.length() > 0)
            {
                FMLLog.bigWarning("Dangerous alternative prefix `{}` for name `{}`, expected `{}` invalid registry invocation/invalid name?", oldPrefix, name, prefix);
                prefix = oldPrefix;
            }
            this.registryName = new ResourceLocation(prefix, name);
            return (T)this;
        }

        //Helper functions
        public final T setRegistryName(ResourceLocation name){ return setRegistryName(name.toString()); }
        public final T setRegistryName(String modID, String name){ return setRegistryName(modID + ":" + name); }
        @Nullable
        public final ResourceLocation getRegistryName()
        {
            if (delegate.name() != null) return delegate.name();
            return registryName != null ? registryName : null;
        }

        @Override
        public final Class<T> getRegistryType() { return (Class<T>) token.getRawType(); };
    }